internal/core/docker_host.go

```diff
@@ -67,9 +84,16 @@ func DefaultGatewayIP() (string, error) {
 //  5. Docker host from the "docker.host" property in the ~/.testcontainers.properties file.
 //  6. Rootless docker socket path.
 //  7. Else, the default Docker socket including schema will be returned.
-func ExtractDockerHost(ctx context.Context) string {
+func ExtractDockerHost(ctx context.Context, callbackChecks ...func(host string) error) string {
```

Comment:
This code is internal, so won't be used by the client code, that's why I decided to go with a variadic arg and not change the APIs that much

Reply:
In which case could we not just force the check vs making it optional, as I couldn't see an place where it was passed in and not the default?

Reply:
It's used in the tests (see comment [here](https://github.com/testcontainers/testcontainers-go/pull/2741/files#diff-a8ee20e4d0804d67968f2493bb05b3ad5265f5ce9456d6d438fbb4f91d18023eR89-R91)):
- https://github.com/testcontainers/testcontainers-go/pull/2741/files#diff-1a24c2a3e6705e9b915bf8fde90c9c3688fc68edc7bafd0528d8db6b3a5fc0dfR6
- https://github.com/testcontainers/testcontainers-go/pull/2741/files#diff-1a24c2a3e6705e9b915bf8fde90c9c3688fc68edc7bafd0528d8db6b3a5fc0dfR447
- https://github.com/testcontainers/testcontainers-go/pull/2741/files#diff-1a24c2a3e6705e9b915bf8fde90c9c3688fc68edc7bafd0528d8db6b3a5fc0dfR90
- https://github.com/testcontainers/testcontainers-go/pull/2741/files#diff-1a24c2a3e6705e9b915bf8fde90c9c3688fc68edc7bafd0528d8db6b3a5fc0dfR63

Reply:
If its just to enable edge case testing in a very limited amount of cases you could use a variable based function.

This would simplify all round, no argument changes and the loop would be simply:
```go
	for _, dockerHostFn := range dockerHostFns {
		dockerHost, err := dockerHostFn(ctx)
		if err != nil {
			continue
		}

		if err = hostCheck(dockerHost); err != nil {
			continue
		}
	}
```

Then add a little test helper:
```go
// customHostCheck overrides hostCheck used by extractDockerHost for
// the duration of the test.
func customHostCheck(t *testing.T, fn func(host string) error) {
	t.Helper()

	old := hostCheck
	hostCheck= fn
	t.Cleanup(func() {
		hostCheck = oldHostCheck
	})
}
```

```go
	t.Run("Docker Host is extracted just once", func(t *testing.T) {
		expected := "/path/to/docker.sock"
		t.Setenv("DOCKER_HOST", expected)
 		customHostCheck(t, testCallbackCheckPassing)

		host := ExtractDockerHost(context.Background())
		require.Equal(t, expected, host)
	})
```
I commonly use that pattern to override `time.Now()` when running tests.

Reply:
Ok, so we create just a default hostCheck (the current default callback check function) which we override in tests, right?

Reply:
@stevenh I addressed this in https://github.com/testcontainers/testcontainers-go/pull/2741/commits/e8eeff38be456f9d531c449dbf5fac6272cff88c

I think we only need to deal with the error 🤔 Because the `ExtractDockerHost` function is internal, I think it's fine to make it return an error, or because testcontainers-go _needs_ a valid socket path, I think we could panic here. 🤔 

I'm not that familiar with how idiomatic is to panic in a library, so I'd ask for advice here.

Reply:
Yer I would absolutely return an error if we can't find a valid docker host.